python/pyarrow/_hdfs.pyx

```diff
@@ -0,0 +1,169 @@
+        if endpoint is not None:
+            self.endpoint = endpoint
+        if driver is not None:
+            self.driver = driver
+        if replication is not None:
+            self.replication = replication
+        if user is not None:
+            self.user = user
+        if buffer_size is not None:
+            self.buffer_size = buffer_size
+        if default_block_size is not None:
+            self.default_block_size = default_block_size
+
+    @staticmethod
+    cdef wrap(CHdfsOptions wrapped):
+        cdef HdfsOptions self = HdfsOptions.__new__(HdfsOptions)
+        self.options = wrapped
+        return self
+
+    cdef inline CHdfsOptions unwrap(self) nogil:
+        return self.options
+
+    @staticmethod
+    def from_uri(uri):
+        cdef CResult[CHdfsOptions] result
+        result = CHdfsOptions.FromUriString(tobytes(uri))
+        return HdfsOptions.wrap(GetResultValue(result))
+
+    @property
+    def endpoint(self):
+        return (
+            frombytes(self.options.connection_config.host),
+            self.options.connection_config.port
+        )
+
+    @endpoint.setter
+    def endpoint(self, value):
+        if not isinstance(value, tuple) or len(value) != 2:
+            raise TypeError('Endpoint must be a tuple of host port pair')
+        self.options.connection_config.host = tobytes(value[0])
+        self.options.connection_config.port = int(value[1])
+
+    @property
+    def driver(self):
+        if self.options.connection_config.driver == HdfsDriver_LIBHDFS3:
+            return 'hdfs3'
+        else:
+            return 'hdfs'
+
+    @driver.setter
+    def driver(self, value):
+        if value == 'hdfs3':
+            self.options.connection_config.driver = HdfsDriver_LIBHDFS3
+        elif value == 'hdfs':
+            self.options.connection_config.driver = HdfsDriver_LIBHDFS
+        else:
+            raise ValueError('Choose either hdfs of hdfs3')
+
+    @property
+    def user(self):
+        return frombytes(self.options.connection_config.user)
+
+    @user.setter
+    def user(self, value):
+        self.options.connection_config.user = tobytes(value)
+
+    @property
+    def replication(self):
+        return self.options.replication
+
+    @replication.setter
+    def replication(self, int value):
+        self.options.replication = value
+
+    @property
+    def buffer_size(self):
+        return self.options.buffer_size
+
+    @buffer_size.setter
+    def buffer_size(self, int value):
+        self.options.buffer_size = value
+
+    @property
+    def default_block_size(self):
+        return self.options.default_block_size
+
+    @default_block_size.setter
+    def default_block_size(self, int value):
+        self.options.default_block_size = value
+
+
+cdef class HadoopFileSystem(FileSystem):
+    """HDFS backed FileSystem implementation
+
+    Parameters
+    ----------
+    options: HdfsOptions, default None
```

Comment:
In the signature this is `options_or_uri`, so also taking an uri string (also the explanation could be updated for that)

Reply:
I think this is also not tested (the ability to directly pass a string)

Reply:
The class method is tested https://github.com/apache/arrow/pull/5990/files#diff-25109fc88b64bdae939cbfd95c669671R530, so the doctoring needs to be updated.

Reply:
Updated.